### PR TITLE
late initializing some xrc fields from the xr

### LINKED ISSUE #2263: Late initialize XRC spec.compositionRef
### What happened?

A composite resource claim (XRC) is a namespaced proxy for a composite resource (XR). Both types have an almost identical schema. When an XRC is created an XR is typically created in response, and the claim's `spec.compositionRef` and `spec.compositionSelector` are propagated to the XR. Currently if you create a claim with only a composition selector, or with neither a reference or a selector (relying on the XRD's default or enforced composition) the claim's composition reference is never updated to reflect the composition selected by the XR.

We mostly "late initialize" a claim from its XR, but when doing so we ignore any fields shared with the XR. This includes the composition reference, selector, etc: https://github.com/crossplane/crossplane/blob/39217a2c44c73d32e16364bf6a8bbcab72910ed3/internal/controller/apiextensions/claim/configurator.go#L149

### How can we reproduce it?

Create a claim with a composition selector and no composition reference (e.g. the GCP claim from the Crossplane 1.1 getting started guide) and observe that the XR has its composition reference set, but the claim never does.

### What environment did it happen in?
Crossplane version: 1.1

### PR DESCRIPTION
**Pull request by RinkiyaKeDad**
Signed-off-by: RinkiyaKeDad <arshsharma461@gmail.com>

### Description of your changes

Followed @negz's suggestion and used the [existing pattern](https://github.com/crossplane/crossplane/blob/06fadd7433bbc9d6b9ba00125f36864e6a2a80c7/internal/controller/apiextensions/claim/configurator.go#L116-L121) of late initializing the claim's spec from the composite's spec.

Fixes #2263 

I have:

- [ ] Read and followed Crossplane's [contribution process].
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: internal/xcrd/schemas.go [hunk 1]
```diff
@@ -29,6 +29,10 @@ const (
 // when translating an XRC into an XR.
 var PropagateClaimSpecProps = []string{"compositionRef", "compositionSelector", "compositionRevisionRef", "compositionUpdatePolicy"}
 
+// PropagateCompositeSpecProps is the list of XR spec properties to propagate
+// when translating an XR into an XRC.
+var PropagateCompositeSpecProps = []string{"compositionRef", "compositionSelector"}
```
Comment (negz) [inline internal/xcrd/schemas.go:34]: I'm pretty sure claims have `compositionRevisionRef` and `compositionUpdatePolicy` too (needs double checking). If so, this list would be identical to `PropagateClaimSpecProps` above. Maybe we could just call it something like `PropagateSpecProps` and use it in both places. 🤔

File: internal/controller/apiextensions/claim/configurator.go [hunk 2]
```diff
@@ -200,13 +200,19 @@ func (c *APIClaimConfigurator) Configure(ctx context.Context, cm resource.Compos
...
-	if err := merge(ucm.Object["spec"], ucp.Object["spec"],
-		withSrcFilter(xcrd.GetPropFields(xcrd.CompositeResourceSpecProps())...)); err != nil {
-		return errors.Wrap(err, errMergeClaimSpec)
+	// We want to propagate the composite's spec to the claim's spec, but
+	// first we must filter out any well-known fields that are unique to
+	// composites. We do this by:
+	// 1. Grabbing a map whose keys represent all well-known composite fields.
+	// 2. Deleting any well-known fields that we want to propagate.
+	// 3. Filtering OUT the remaining map keys from the composite's spec so
+	// 	  that we end up adding only the well-known fields to the claim's spec.
```
Comment (negz) [inline internal/controller/apiextensions/claim/configurator.go:209]: Nit: Some superfluous leading white space here.

File: internal/controller/apiextensions/claim/configurator.go [hunk 3]
```diff
@@ -200,13 +200,19 @@ func (c *APIClaimConfigurator) Configure(ctx context.Context, cm resource.Compos
...
+	wellKnownCompositeFields := xcrd.CompositeResourceSpecProps()
+	for _, field := range xcrd.PropagateCompositeSpecProps {
+		delete(wellKnownCompositeFields, field)
 	}
+	compositeSpecFilter := xcrd.GetPropFields(wellKnownCompositeFields)
+	ucm.Object["spec"] = filter(ucp.Object["spec"].(map[string]interface{}), compositeSpecFilter...)
```
Comment (negz) [inline internal/controller/apiextensions/claim/configurator.go:215]: I think we need to keep the `merge` pattern here, so I _think_ we'd just need to change `withSrcFilter` to `withSrcFilter(compositeSpecFilter)`.

**Review by negz — COMMENTED** [review]
Thanks for opening this @RinkiyaKeDad! I think a few changes are needed. Once you've addressed the comments can you please edit the PR description to provide a little more detail about what this change does, and how it was tested.

**Comment by RinkiyaKeDad** [comment]
Hey @negz, thanks for the review! I'm actually facing some trouble trying to set up a development environment for Crossplane. Will test the changes and update the PR once that is sorted out.

[Slack thread](https://crossplane.slack.com/archives/CEF5N8X08/p1635344265053100) for context.

**Comment by negz** [comment]
>  I'm actually facing some trouble trying to set up a development environment for Crossplane. Will test the changes and update the PR once that is sorted out.

Got it! Responding here just in case you don't see my response on Slack - I think the issue was that you're running Crossplane outside a container (e.g. via `go run` or `make run`) and no `/cache` directory exists for it to use as a package cache. `mkdir /tmp/cache && go run main.go start -d -c /tmp/cache` should fix that.

File: internal/xcrd/schemas.go (hunk 1 above)
Comment (RinkiyaKeDad) [reply internal/xcrd/schemas.go:34]: > needs double checking

I only see `CompositeResourceDefinition` in the API Docs here - https://doc.crds.dev/github.com/crossplane/crossplane/apiextensions.crossplane.io/CompositeResourceDefinition/v1@v1.5.0

And the "Create CompositeResourceDefinition" example [here](https://crossplane.io/docs/v1.5/getting-started/create-configuration.html#create-compositeresourcedefinition) uses the same YAML to define the schema for the XRC and the XR so I believe you were right :) 

Will update to using a single `PropagateSpecProps` variable for both.

Comment (RinkiyaKeDad) [reply internal/xcrd/schemas.go:34]: Triple checked by running `kubectl explain postgresqlinstance.spec`:

```
KIND:     PostgreSQLInstance
VERSION:  database.example.org/v1alpha1

RESOURCE: spec <Object>

DESCRIPTION:
     <empty>

FIELDS:
   compositionRef	<Object>

   compositionRevisionRef	<Object>

   compositionSelector	<Object>

   compositionUpdatePolicy	<string>

   parameters	<Object> -required-

   resourceRef	<Object>

   writeConnectionSecretToRef	<Object>
```

**Comment by negz** [comment]
@RinkiyaKeDad Thank you, this looks great! Could you please update the checklist in the PR description and also add some details on how you've tested this before I merge. Speaking of tests, are there any obvious unit tests covering this functionality we could update or expand?

**Comment by RinkiyaKeDad** [comment]
Hi @negz sorry I didn't get time to get back to this. I will update after testing it locally and seeing if we can add unit tests for it by next week :) Thank you for reviewing!

**Comment by negz** [comment]
Friendly ping. This still looks good - just waiting on info on how this was tested (and unit tests if possible.) :)

**Comment by RinkiyaKeDad** [comment]
Hey @negz, thanks for the nudge! The error I was facing [while testing](https://crossplane.io/docs/v1.5/getting-started/install-configure.html#install-configuration-package-2) was after running: 
```
kubectl crossplane install configuration registry.upbound.io/xp/getting-started-with-gcp:v1.5.1
```

```
2021-11-30T14:11:02.359+0530    DEBUG   crossplane.controller-runtime.manager.events    Warning {"object": {"kind":"ConfigurationRevision","name":"xp-getting-started-with-gcp-d4bf3c1cc96b","uid":"dbc26804-2a10-48f5-a60a-fc4c4bca7235","apiVersion":"pkg.crossplane.io/v1","resourceVersion":"1841"}, "reason": "LintPackage", "message": "package is not compatible with Crossplane version (): Invalid Semantic Version"}
```

You suggested on Slack to use [spec.ignoreCrossplaneConstraints](https://crossplane.io/docs/v1.5/concepts/packages.html#specignorecrossplaneconstraints) to fix this, but I'm not able to figure out how to change the spec of `registry.upbound.io/xp/getting-started-with-gcp:v1.5.1` while testing? Is there someplace I can get the YAML for this so that I can edit the spec? Thanks!

**Comment by negz** [comment]
> Is there someplace I can get the YAML for this so that I can edit the spec? Thanks!

There is. For context, when you run `kubectl crossplane install configuration ...` a new `Configuration` resource is created in the cluster. You can find it by running `kubectl get configuration.pkg` and then you should be able to edit its spec by running `kubectl edit configuration.pkg xp-getting-started-with-gcp`. When you do, the yaml should look like this:

```yaml
# $ kubectl get -o yaml configuration.pkg xp-getting-started-with-gcp
apiVersion: pkg.crossplane.io/v1
kind: Configuration
metadata:
  creationTimestamp: "2021-11-30T21:58:25Z"
  generation: 1
  name: xp-getting-started-with-gcp
  resourceVersion: "1784"
  uid: 50e6a89a-4e6e-4928-ace4-73b94fece928
spec:
  # Set this to 'true'.
  ignoreCrossplaneConstraints: false
  package: registry.upbound.io/xp/getting-started-with-gcp:v1.5.1
  packagePullPolicy: IfNotPresent
  revisionActivationPolicy: Automatic
  revisionHistoryLimit: 0
  skipDependencyResolution: false
```

Note that you may need to do the same thing for provider-gcp (which is a dependency of the getting started configuration) by running `kubectl edit provider.pkg`.

**Comment by RinkiyaKeDad** [comment]
Hey @negz sorry for the late response. I tried what you suggested and looks like the fix in this PR is working! These are the results I got after [running](https://crossplane.io/docs/v1.5/getting-started/provision-infrastructure.html#claim-your-infrastructure) `kubectl apply -f https://raw.githubusercontent.com/crossplane/crossplane/release-1.5/docs/snippets/compose/claim-gcp.yaml`:


1. `kubectl get composite my-db-ggvz9 -o yaml`:

```
apiVersion: database.example.org/v1alpha1
kind: XPostgreSQLInstance
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"database.example.org/v1alpha1","kind":"PostgreSQLInstance","metadata":{"annotations":{},"name":"my-db","namespace":"default"},"spec":{"compositionSelector":{"matchLabels":{"provider":"gcp"}},"parameters":{"storageGB":20},"writeConnectionSecretToRef":{"name":"db-conn"}}}
  creationTimestamp: "2021-12-09T07:55:34Z"
  generateName: my-db-
  generation: 4
  labels:
    crossplane.io/claim-name: my-db
    crossplane.io/claim-namespace: default
    crossplane.io/composite: my-db-ggvz9
  name: my-db-ggvz9
  resourceVersion: "7221"
  uid: 37bb5f02-df7c-4910-be3c-afac10d9eccb
spec:
  claimRef:
    apiVersion: database.example.org/v1alpha1
    kind: PostgreSQLInstance
    name: my-db
    namespace: default
  compositionRef:
    name: xpostgresqlinstances.gcp.database.example.org
  compositionSelector:
    matchLabels:
      provider: gcp
  compositionUpdatePolicy: Automatic
  parameters:
    storageGB: 20
  resourceRefs:
  - apiVersion: database.gcp.crossplane.io/v1beta1
    kind: CloudSQLInstance
    name: my-db-ggvz9-snfhm
  writeConnectionSecretToRef:
    name: 37bb5f02-df7c-4910-be3c-afac10d9eccb
    namespace: crossplane-system
status:
  conditions:
  - lastTransitionTime: "2021-12-09T07:55:34Z"
    reason: Creating
    status: "False"
    type: Ready
  connectionDetails:
    lastPublishedTime: "2021-12-09T07:57:04Z"
```

2. `kubectl get claim my-db -o yaml`:

```
apiVersion: database.example.org/v1alpha1
kind: PostgreSQLInstance
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"database.example.org/v1alpha1","kind":"PostgreSQLInstance","metadata":{"annotations":{},"name":"my-db","namespace":"default"},"spec":{"compositionSelector":{"matchLabels":{"provider":"gcp"}},"parameters":{"storageGB":20},"writeConnectionSecretToRef":{"name":"db-conn"}}}
  creationTimestamp: "2021-12-09T07:55:34Z"
  finalizers:
  - finalizer.apiextensions.crossplane.io
  generation: 3
  name: my-db
  namespace: default
  resourceVersion: "6935"
  uid: 185ce6fb-b600-423a-bc3c-c7e9b9241cec
spec:
  compositionRef:
    name: xpostgresqlinstances.gcp.database.example.org
  compositionSelector:
    matchLabels:
      provider: gcp
  compositionUpdatePolicy: Automatic
  parameters:
    storageGB: 20
  resourceRef:
    apiVersion: database.example.org/v1alpha1
    kind: XPostgreSQLInstance
    name: my-db-ggvz9
  writeConnectionSecretToRef:
    name: db-conn
status:
  conditions:
  - lastTransitionTime: "2021-12-09T07:55:34Z"
    reason: Composite resource claim is waiting for composite resource to become Ready
    status: "False"
    type: Ready
```



Both the XR and the XRC now have `compositionRef` and `compositionSelector` set. Could you please point me to some similar unit tests I could look at before adding the ones for this change?

Thanks for your patience!

**Comment by negz** [comment]
It's Crossplane v1.6 code freeze today, and I'd really like this feature to land so I'm going to merge this. The code is covered by tests already, and it sounds like you've thoroughly manually tested. I'll take a quick look after merging to see if there are any easy unit tests I could add. Thanks so much for working on this!

**Review by negz — APPROVED** [review]

**Pull request merged**